lib/compress/zstd_fast.c

```diff
@@ -73,26 +98,62 @@ size_t ZSTD_compressBlock_fast_generic(
         U32 const current = (U32)(ip-base);
         U32 const matchIndex = hashTable[h];
         const BYTE* match = base + matchIndex;
+        const U32 repIndex = current + 1 - offset_1;
+        const BYTE* repMatch = (dictMode == ZSTD_dictMatchState
+                            && repIndex < prefixLowestIndex) ?
+                               dictBase + (repIndex - dictIndexDelta) :
+                               base + repIndex;
         hashTable[h] = current;   /* update hash table */
 
-        if ((offset_1 > 0) & (MEM_read32(ip+1-offset_1) == MEM_read32(ip+1))) {
+        if (dictMode == ZSTD_dictMatchState
+            && ((U32)((prefixLowestIndex-1) - repIndex) >= 3 /* intentional underflow */)
```

Comment:
I *think* it is impossible to have invalid repcodes when we are using a cdict with `forceWindow == 0`, since we check them during cdict construction. In which case we could turn this into an assert, and add some extra asserts to the beginning, and remove this check, which might get us some speed. If it doesn't we should leave it though.

Reply:
This check isn't to see if it's out of bounds, it's to see whether the 4 byte read is split across the end of the dictionary and the beginning of the prefix.